lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -995,6 +1016,41 @@ void LifetimeChecker::handleStoreUse(unsigned UseID) {
     }
   }
 
+  // Check if we're in a cross-module struct initializer that uses RootSelf rather
+  // than DelegatingSelf for Swift 4 compatibility. We look for a problem case by
+  // seeing if there are any assignments to individual fields that might be
+  // initializations; that is, that they're not dominated by `self = other`.
+  auto isFullValueAssignment = [this](const SILInstruction *inst) -> bool {
+    SILType ty;
+    if (auto *assign = dyn_cast<AssignInst>(inst))
+      ty = assign->getSrc()->getType();
+    else if (auto *copyAddr = dyn_cast<CopyAddrInst>(inst))
+      ty = copyAddr->getSrc()->getType().getObjectType();
+    return ty && ty == TheMemory.MemorySILType;
+  };
+
+  if (!isFullyInitialized && WantsCrossModuleStructInitializerDiagnostic &&
+      !isFullValueAssignment(Use.Inst)) {
+    // Deliberately don't check shouldEmitError here; we're using DI to approximate
+    // whether this would be a valid delegating initializer, but the error when it
+    // /is/ a delegating initializer won't be path-sensitive.
+
+    Type selfTy;
+    SILLocation fnLoc = TheMemory.getFunction().getLocation();
+    if (auto *ctor = fnLoc.getAsASTNode<ConstructorDecl>())
+      selfTy = ctor->getImplicitSelfDecl()->getType()->getInOutObjectType();
```

Comment:
Why isn't this always just TheMemory.getType()?

Reply:
This way preserves (some) sugar, which is nice.

Reply:
If there's a better way to get at the decl or sugared type I'd be glad to get rid of the ugly reaching up into the function.

Reply:
How about `selfTy.getAsNominalOrBoundGenericNominal()->getDeclaredInterfaceType()`?

Reply:
That doesn't handle constrained extensions (see test case).

Reply:
I don't think that matters, instead you can use getDeclaredType() to print the type without any generic arguments at all

Reply:
Well, I wanted the generic arguments, but if you think it's better without them then that would make things simpler.